### PR TITLE
udp: Fix `UdpFramed` with regards to `Decode`

### PR DESCRIPTION
**Pull request by JohnDoneth**
Duplicate of #1444 but for master instead of the v0.1.x branch.

Should (fix) #1443 on master.

`UdpFramed` now attempts to continue decoding past the first discrete item if there is more data in the read buffer. I added a test for this that sends a packet with 3 lines and tests for the successful decoding of each item in turn.

### CODE REVIEW

**Comment by LucioFranco** [comment]
@JohnDoneth thanks for doing this, I'm gonna review #1444 then when that commit is approved we can port the fix into master.

**Comment by ipetkov** [comment]
@LucioFranco I think this is ready for review now that #1444 has been merged!

**Comment by LucioFranco** [comment]
Yeah, we need to update this to the new method done in #1444 that uses decode_eof instead of just plain decode.

File: tokio-net/src/udp/frame.rs [hunk 1]
```diff
@@ -45,25 +47,38 @@ impl<C: Decoder + Unpin> Stream for UdpFramed<C> {
...
-            // Read into the buffer without having to initialize the memory.
-            let res = ready!(Pin::new(&mut pin.socket).poll_recv_from_priv(cx, pin.rd.bytes_mut()));
-            let (n, addr) = res?;
-            pin.rd.advance_mut(n);
-            (n, addr)
-        };
+        loop {
+            // Are there are still bytes left in the read buffer to decode?
+            if pin.is_readable {
+                if let Some(frame) = pin.codec.decode(&mut pin.rd)? {
```
Comment (LucioFranco) [inline tokio-net/src/udp/frame.rs]: so I think we want this to be decode_eof for it to work properly.

Comment (Ralith) [reply tokio-net/src/udp/frame.rs]: Shouldn't that only be the last call, rather than all of them?

**Comment by JohnDoneth** [comment]
I am working on this again. Currently stuck on this API change where `BytesMut` returns a slice of possibly uninitialized bytes.
```
error[E0308]: mismatched types
  --> tokio-util\src\udp\frame.rs:72:79
   |
72 |                 let res = ready!(Pin::new(&mut pin.socket).poll_recv_from(cx, pin.rd.bytes_mut()));
   |                                                                               ^^^^^^^^^^^^^^^^^^ expected u8, found union `std::mem::MaybeUninit`
   |
   = note: expected type `&mut [u8]`
              found type `&mut [std::mem::MaybeUninit<u8>]`
```

File: tokio-net/src/udp/frame.rs (hunk 1 above)
Comment (JohnDoneth) [reply tokio-net/src/udp/frame.rs]: Done!

**Comment by Darksonn** [comment]
I was wondering what the status on this is?

**Comment by JohnDoneth** [comment]
@Darksonn It was ready to be reviewed. It now needs to be caught up to master again

**Comment by Darksonn** [comment]
Sorry about that. If you merge master again I'll make sure someone actually reviews it :)

**Comment by LucioFranco** [comment]
@JohnDoneth sorry about the delay, but anyway you could get this rebased and we can get it merged?

**Comment by JohnDoneth** [comment]
I will put it on my todo list!

**Comment by LucioFranco** [comment]
@JohnDoneth looks like some clippy and rustfmt failures?

**Comment by JohnDoneth** [comment]
I think it might be CI? `cargo clippy` doesn't produce any output locally and `cargo fmt` leaves the project unchanged.

**Comment by Darksonn** [comment]
The clippy error is the following. You probably forgot an `--all-features` on the clippy invocation.
```
error: transmute from a reference to a reference
  --> tokio-util/src/udp/frame.rs:71:36
   |
71 |                 let b: &mut [u8] = std::mem::transmute(pin.rd.bytes_mut());
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&mut *(pin.rd.bytes_mut() as *mut [std::mem::MaybeUninit<u8>] as *mut [u8])`
   |
   = note: `-D clippy::transmute-ptr-to-ptr` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#transmute_ptr_to_ptr
```
As for rustfmt, it's pretty confusing but you have to use the following command in Tokio's repositories. This has to do with our extensive use of features together with the use of macros to enable/disable them, which break rustfmt in a way that means it cannot find all the files.
```
rustfmt --edition 2018 $(find . -name '*.rs' -print)
```
Check out [our contributing guide](https://github.com/tokio-rs/tokio/blob/master/CONTRIBUTING.md#cargo-commands) for more info.

**Comment by JohnDoneth** [comment]
@Darksonn Thanks! That was the problem.

I believe this PR is ready now.

ping @LucioFranco

**Comment by LucioFranco** [comment]
@JohnDoneth thanks a ton for following up!

**Comment by kurnevsky** [comment]
Looks like this change makes decoding errors useless. If we return error from `decode` function then we will try to parse the same bytes from the same udp packet and will get the same error, and so on ad infinitum. @JohnDoneth was it expected? Should we put decoding errors to the `Item` type now?

**Comment by Darksonn** [comment]
Ah, that sounds like a mistake. Consider opening a PR to fix that.

**Comment by kurnevsky** [comment]
Looked through the implementation once again and realized that we have to clear the buffer when packet was parsed, and then have to handle empty input as None. So we could do the same when error happens actually.

**Comment by kurnevsky** [comment]
But then we won't be able to distinguish zero sized packets from completely consumed packets. Probably we should check for buffer size here https://github.com/tokio-rs/tokio/blob/master/tokio-util/src/udp/frame.rs#L63 ? And then do something similar for errors?

**Pull request merged**